addon/services/i18n.js

```diff
@@ -19,11 +19,26 @@ export default Parent.extend(Ember.Evented, {
   locales: computed(getLocales),
 
   // @public
+  // Whether to allow overriding the locale for individual translations.
+  allowLocaleOverride: undefined,
+
+  // @public
   //
   // Returns the translation `key` interpolated with `data`
   // in the current `locale`.
   t: function(key, data = {}) {
-    const locale = this.get('_locale');
+    let locale = this.get('_locale');
+    if (data.locale && this.get('allowLocaleOverride') !== false) {
+      if (this.get('allowLocaleOverride') === undefined) {
+        Ember.warn(
+          'Specifying an interpolation key named `locale` now overrides the target locale for translation. To silence this warning, set ENV.i18n.allowLocaleOverride to true. To disable this behavior and treat `locale` as a normal interpolation key, set ENV.i18n.allowLocaleOverride to false.',
+          { id: 'ember-i18n.locale-override' }
+        );
+      }
+      locale = new Locale(data.locale, this.container);
```

Comment:
This part seems inefficient, but addressing it would probably require a larger change (for instance, locales could be cached on the service when they are initialized).